### PR TITLE
Add doc topic for unmanaged calling conventions

### PR DESCRIPTION
**Pull request by jkotas**
## Summary

Add doc topic for unmanaged calling conventions. Addresses feedback from https://github.com/dotnet/runtime/issues/90462#issuecomment-1676049957

Other miscellaneous interop doc improvements:
- Rename article title to match the content
- Replace link to PInvoke.net wiki with a link to CsWin32. CsWin32 is more modern and precise.
- Update Linux and macOS samples to follow best practices

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/dependency-loading/default-probing.md](https://github.com/dotnet/docs/blob/33e9434e20155b56b476f2194a383e18847b168f/docs/core/dependency-loading/default-probing.md) | [Default probing](https://review.learn.microsoft.com/en-us/dotnet/core/dependency-loading/default-probing?branch=pr-en-us-36768) |
| [docs/standard/native-interop/calling-conventions.md](https://github.com/dotnet/docs/blob/33e9434e20155b56b476f2194a383e18847b168f/docs/standard/native-interop/calling-conventions.md) | [docs/standard/native-interop/calling-conventions](https://review.learn.microsoft.com/en-us/dotnet/standard/native-interop/calling-conventions?branch=pr-en-us-36768) |
| [docs/standard/native-interop/native-library-loading.md](https://github.com/dotnet/docs/blob/33e9434e20155b56b476f2194a383e18847b168f/docs/standard/native-interop/native-library-loading.md) | [docs/standard/native-interop/native-library-loading](https://review.learn.microsoft.com/en-us/dotnet/standard/native-interop/native-library-loading?branch=pr-en-us-36768) |
| [docs/standard/native-interop/pinvoke.md](https://github.com/dotnet/docs/blob/33e9434e20155b56b476f2194a383e18847b168f/docs/standard/native-interop/pinvoke.md) | [Platform Invoke (P/Invoke)](https://review.learn.microsoft.com/en-us/dotnet/standard/native-interop/pinvoke?branch=pr-en-us-36768) |

### CODE REVIEW

**Comment by jkotas** [comment]
Once this goes through, I plan to submit a matching update in dotnet-api-docs and update default calling convention mentions to point to this topic.

File: .openpublishing.redirection.standard.json [hunk 1]
```diff
@@ -808,6 +808,10 @@
             "source_path_from_root": "/docs/standard/using-linq.md",
             "redirect_url": "/dotnet/standard/linq/",
             "redirect_document_id": true
+        },
+        {
+            "source_path_from_root": "/docs/standard/native-interop/cross-platform.md",
+            "redirect_url": "/dotnet/standard/native-interop/pinvoke"
```
Comment (jkotas) [inline .openpublishing.redirection.standard.json:814]: Do I need to set `redirect_document_id` here?

File: docs/standard/native-interop/calling-conventions.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
+---
+title: Unmanaged calling conventions
+description: Learn how to specify unmanaged calling conventions in .NET
+author: jkotas
+ms.date: 08/19/2023
+---
+
+# Unmanaged calling conventions
+
+[Calling conventions](https://en.wikipedia.org/wiki/Calling_convention) desribe low-level details for how method arguments and return values
```
Comment (AaronRobinsonMSFT) [inline docs/standard/native-interop/calling-conventions.md:10]:
```suggestion
[Calling conventions](https://en.wikipedia.org/wiki/Calling_convention) describe low-level details for how method arguments and return values
```

File: docs/standard/native-interop/calling-conventions.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+are passed between the caller and the called method.
+
+It is important that the unmanaged calling convention declared in P/Invoke declaration matches the unmanaged calling convention
+used by the native implementation. Unmanaged calling convention mismatches lead to data corruptions and fatal crashes that require
+low-level debugging skills to diagnose.
+
+## Platform default calling convention
+
+Most platforms use one canonical calling conventions and explicitly specified calling convention is unnecessary in most cases.
```
Comment (AaronRobinsonMSFT) [inline docs/standard/native-interop/calling-conventions.md:19]:
```suggestion
Most platforms use one canonical calling convention and an explicitly specified calling convention is unnecessary in most cases.
```

File: docs/standard/native-interop/calling-conventions.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+For x86 architecture, the default calling convention is platform-specific. StdCall is the default calling convention on Windows x86
```
Comment (AaronRobinsonMSFT) [inline docs/standard/native-interop/calling-conventions.md:21]:
```suggestion
For the x86 architecture, the default calling convention is platform-specific. Stdcall ("standard call") is the default calling convention on Windows x86
```

File: docs/standard/native-interop/calling-conventions.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+and it is used by most Win32 APIs. Cdecl is the default calling convention on Linux x86. Windows ports of open source libraries that
+originated on Unix often use Cdecl calling convention even on Windows x86 and it is necessary to explicitly specifify the Cdecl calling
```
Comment (AaronRobinsonMSFT) [inline docs/standard/native-interop/calling-conventions.md:23]:
```suggestion
originated on Unix often use the Cdecl calling convention even on Windows x86 and it is necessary to explicitly specify the Cdecl calling
```

File: docs/standard/native-interop/calling-conventions.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+convention in P/Invoke declarations for interop with these libraries.
+
+For non-x86 architectures, both StdCall and Cdecl calling conventions are treated as the canonical platform default calling convention.
+
+## Specifying calling conventions in managed P/Invoke declarations
+
+The calling conventions are specific by types in System.Runtime.CompilerServices namespace or their combinations:
```
Comment (AaronRobinsonMSFT) [inline docs/standard/native-interop/calling-conventions.md:30]:
```suggestion
The calling conventions are specific by types in the System.Runtime.CompilerServices namespace or their combinations:
```

File: docs/standard/native-interop/calling-conventions.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+## Platform default calling convention
+
+Most platforms use one canonical calling conventions and explicitly specified calling convention is unnecessary in most cases.
+
+For x86 architecture, the default calling convention is platform-specific. StdCall is the default calling convention on Windows x86
+and it is used by most Win32 APIs. Cdecl is the default calling convention on Linux x86. Windows ports of open source libraries that
+originated on Unix often use Cdecl calling convention even on Windows x86 and it is necessary to explicitly specifify the Cdecl calling
+convention in P/Invoke declarations for interop with these libraries.
+
+For non-x86 architectures, both StdCall and Cdecl calling conventions are treated as the canonical platform default calling convention.
```
Comment (AaronRobinsonMSFT) [inline docs/standard/native-interop/calling-conventions.md:26]:
```suggestion
For non-x86 architectures, both Stdcall and Cdecl calling conventions are treated as the canonical platform default calling convention.
```

File: docs/standard/native-interop/calling-conventions.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+static unsafe int NativeCallback(void* context);
+
+// Method returning function pointer with combination of Cdecl and MemberFunction calling conventions.
+static unsafe delegate* unmanaged[Cdecl, MemberFunction]<int> GetHandler();
+```
+
+## Specifying calling conventions in unmanaged C/C++ code
+
+The default calling convention in unmanaged code is typically specified project-wide via C/C++ compiler options. For individual methods,
+the calling conventions can be specified using the `__stdcall` and `__Cdecl` compiler intrinsics under Microsoft Visual C++, and by using
```
Comment (AaronRobinsonMSFT) [inline docs/standard/native-interop/calling-conventions.md:61]:
```suggestion
the calling conventions can be specified using the `__stdcall` and `__cdecl` compiler intrinsics under Microsoft Visual C++, and by using
```

File: docs/standard/native-interop/calling-conventions.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+the `__attribute__((stdcall))` and `__attribute__((Cdecl))` compiler intrinsics under GCC and Clang.
```
Comment (AaronRobinsonMSFT) [inline docs/standard/native-interop/calling-conventions.md:62]:
```suggestion
the `__attribute__((stdcall))` and `__attribute__((cdecl))` compiler intrinsics under GCC and Clang.
```

File: docs/standard/native-interop/calling-conventions.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+## Specifying calling conventions in earlier .NET versions
+
+.NET Framework and .NET versions prior to .NET 5 are limited to subset of calling conventions that can described by [CallingConvention](xref:System.Runtime.InteropServices.CallingConvention) enumeration.
```
Comment (AaronRobinsonMSFT) [inline docs/standard/native-interop/calling-conventions.md:66]:
```suggestion
.NET Framework and .NET versions prior to .NET 5 are limited to a subset of calling conventions that can described by [CallingConvention](xref:System.Runtime.InteropServices.CallingConvention) enumeration.
```

File: docs/standard/native-interop/calling-conventions.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+Most platforms use one canonical calling conventions and explicitly specified calling convention is unnecessary in most cases.
+
+For x86 architecture, the default calling convention is platform-specific. StdCall is the default calling convention on Windows x86
+and it is used by most Win32 APIs. Cdecl is the default calling convention on Linux x86. Windows ports of open source libraries that
+originated on Unix often use Cdecl calling convention even on Windows x86 and it is necessary to explicitly specifify the Cdecl calling
+convention in P/Invoke declarations for interop with these libraries.
+
+For non-x86 architectures, both StdCall and Cdecl calling conventions are treated as the canonical platform default calling convention.
+
+## Specifying calling conventions in managed P/Invoke declarations
```
Comment (AaronRobinsonMSFT) [inline docs/standard/native-interop/calling-conventions.md:28]: Do we want to mention the `CallingConvention.Winapi` option?

File: docs/standard/native-interop/calling-conventions.md (hunk 4 above)
Comment (AaronRobinsonMSFT) [reply docs/standard/native-interop/calling-conventions.md:21]: Do we want to mention the name mangling between stdcall and cdecl? For example `Foo@4` vs `_Foo`. That tends to be something people often get caught up on.

File: docs/standard/native-interop/calling-conventions.md (hunk 11 above)
Comment (jkotas) [reply docs/standard/native-interop/calling-conventions.md:28]: `CallingConvention.Winapi` is no-op. It is same as default; people do not really need to know that it exists. I think it is enough for it to be mentioned in https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.callingconvention?view=net-7.0#fields . 

I plan to add link from CallingConvention.Winapi docs to "Platform default calling convention" paragraph in this doc as a follow up.

File: docs/standard/native-interop/calling-conventions.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+For the x86 architecture, the default calling convention is platform-specific. Stdcall ("standard call") is the default calling convention on Windows x86
+and it is used by most Win32 APIs. Cdecl is the default calling convention on Linux x86. Windows ports of open source libraries that
+originated on Unix often use the Cdecl calling convention even on Windows x86 and it is necessary to explicitly specify the Cdecl calling
+convention in P/Invoke declarations for interop with these libraries.
+
+For non-x86 architectures, both Stdcall and Cdecl calling conventions are treated as the canonical platform default calling convention.
+
+## Specifying calling conventions in managed P/Invoke declarations
+
+The calling conventions are specific by types in the System.Runtime.CompilerServices namespace or their combinations:
```
Comment (jkotas) [inline docs/standard/native-interop/calling-conventions.md:30]:
```suggestion
The calling conventions are specific by types in the `System.Runtime.CompilerServices` namespace or their combinations:
```

File: docs/standard/native-interop/calling-conventions.md (hunk 4 above)
Comment (jkotas) [reply docs/standard/native-interop/calling-conventions.md]: I think it is enough to just give people a few hints that calling conventions are a problem on Windows x86. This is not meant to be Windows x86 calling convention troubleshooting doc. I would not mind reducing the Windows x86 specific coverage even more.

Both wikipedia and stackoverflow questions have coverage for the Windows x86 calling convention name mangling issues.

File: docs/standard/native-interop/calling-conventions.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+// Method returning function pointer with combination of Cdecl and MemberFunction calling conventions.
+static unsafe delegate* unmanaged[Cdecl, MemberFunction]<int> GetHandler();
+```
+
+## Specifying calling conventions in unmanaged C/C++ code
+
+The default calling convention in unmanaged code is typically specified project-wide via C/C++ compiler options. For individual methods,
+the calling conventions can be specified using the `__stdcall` and `__cdecl` compiler intrinsics under Microsoft Visual C++, and by using
+the `__attribute__((stdcall))` and `__attribute__((cdecl))` compiler intrinsics under GCC and Clang.
+
```
Comment (jkotas) [inline docs/standard/native-interop/calling-conventions.md:63]:
```suggestion
```
This does not look high-enough value to include in this topic.

File: .openpublishing.redirection.standard.json (hunk 1 above)
Comment (gewarren) [inline .openpublishing.redirection.standard.json:814]: Shouldn't this point to native-library-loading?

File: docs/standard/native-interop/pinvoke.md [hunk 14]
```diff
@@ -60,6 +60,6 @@ Both of the previous examples depend on parameters, and in both cases, the param
 
 ## More resources
 
-- [PInvoke.net wiki](https://www.pinvoke.net/) an excellent Wiki with information on common Windows APIs and how to call them.
+- [C#/Win32 P/Invoke Source Generator](https://github.com/microsoft/CsWin32/) automatically generates definitions for Windows APIs.
```
Comment (gewarren) [inline docs/standard/native-interop/pinvoke.md:63]:
```suggestion
- [C#/Win32 P/Invoke source generator](https://github.com/microsoft/CsWin32/) automatically generates definitions for Windows APIs.
```

File: docs/standard/native-interop/calling-conventions.md [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+author: jkotas
+ms.date: 08/19/2023
+---
+
+# Unmanaged calling conventions
+
+[Calling conventions](https://en.wikipedia.org/wiki/Calling_convention) describe low-level details for how method arguments and return values
+are passed between the caller and the called method.
+
+It is important that the unmanaged calling convention declared in P/Invoke declaration matches the unmanaged calling convention
```
Comment (gewarren) [inline docs/standard/native-interop/calling-conventions.md:13]:
```suggestion
It's important that the unmanaged calling convention declared in a P/Invoke declaration matches the unmanaged calling convention
```

File: docs/standard/native-interop/calling-conventions.md [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+used by the native implementation. Unmanaged calling convention mismatches lead to data corruptions and fatal crashes that require
```
Comment (gewarren) [inline docs/standard/native-interop/calling-conventions.md:14]:
```suggestion
used by the native implementation. Mismatches in unmanaged calling conventions lead to data corruptions and fatal crashes that require
```

File: docs/standard/native-interop/calling-conventions.md [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+low-level debugging skills to diagnose.
+
+## Platform default calling convention
+
+Most platforms use one canonical calling convention and an explicitly specified calling convention is unnecessary in most cases.
+
+For the x86 architecture, the default calling convention is platform-specific. Stdcall ("standard call") is the default calling convention on Windows x86
```
Comment (gewarren) [inline docs/standard/native-interop/calling-conventions.md:21]:
```suggestion
For the x86 architecture, the default calling convention is platform specific. `Stdcall` ("standard call") is the default calling convention on Windows x86
```

File: docs/standard/native-interop/calling-conventions.md [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+and it is used by most Win32 APIs. Cdecl is the default calling convention on Linux x86. Windows ports of open source libraries that
```
Comment (gewarren) [inline docs/standard/native-interop/calling-conventions.md:22]:
```suggestion
and it is used by most Win32 APIs. `Cdecl` is the default calling convention on Linux x86. Windows ports of open-source libraries that
```

File: docs/standard/native-interop/calling-conventions.md [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+originated on Unix often use the Cdecl calling convention even on Windows x86 and it is necessary to explicitly specify the Cdecl calling
```
Comment (gewarren) [inline docs/standard/native-interop/calling-conventions.md:23]:
```suggestion
originated on Unix often use the `Cdecl` calling convention even on Windows x86. It's necessary to explicitly specify the `Cdecl` calling
```

File: docs/standard/native-interop/calling-conventions.md [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+convention in P/Invoke declarations for interop with these libraries.
+
+For non-x86 architectures, both Stdcall and Cdecl calling conventions are treated as the canonical platform default calling convention.
+
+## Specifying calling conventions in managed P/Invoke declarations
+
+The calling conventions are specific by types in the `System.Runtime.CompilerServices` namespace or their combinations:
```
Comment (gewarren) [inline docs/standard/native-interop/calling-conventions.md:30]:
```suggestion
The calling conventions are specified by types in the `System.Runtime.CompilerServices` namespace or their combinations:
```

File: docs/standard/native-interop/calling-conventions.md [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+- [CallConvSuppressGCTransition](xref:System.Runtime.CompilerServices.CallConvSuppressGCTransition)
+- [CallConvThiscall](xref:System.Runtime.CompilerServices.CallConvThiscall)
+
+Examples of explicitly specified calling conventions:
+
+```csharp
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+// P/Invoke declaration using SuppressGCTransition calling convention
```
Comment (gewarren) [inline docs/standard/native-interop/calling-conventions.md:45]:
```suggestion
// P/Invoke declaration using SuppressGCTransition calling convention.
```

**Review by gewarren — APPROVED** [review]

File: .openpublishing.redirection.standard.json (hunk 1 above)
Comment (jkotas) [inline .openpublishing.redirection.standard.json:814]:
```suggestion
            "redirect_url": "/dotnet/standard/native-interop/pinvoke/native-library-loading.md",
            "redirect_document_id": true
```

Comment (jkotas) [reply .openpublishing.redirection.standard.json]: Fixed. I was not sure whether it is better to make it point to the renamed doc or the general doc.

File: .openpublishing.redirection.standard.json [hunk 22]
```diff
@@ -808,6 +808,11 @@
             "source_path_from_root": "/docs/standard/using-linq.md",
             "redirect_url": "/dotnet/standard/linq/",
             "redirect_document_id": true
+        },
+        {
+            "source_path_from_root": "/docs/standard/native-interop/cross-platform.md",
+            "redirect_url": "/dotnet/standard/native-interop/pinvoke/native-library-loading.md",
```
Comment (jkotas) [inline .openpublishing.redirection.standard.json:814]:
```suggestion
            "redirect_url": "/dotnet/standard/native-interop/pinvoke/native-library-loading",
```

File: .openpublishing.redirection.standard.json [hunk 23]
```diff
@@ -808,6 +808,11 @@
             "source_path_from_root": "/docs/standard/using-linq.md",
             "redirect_url": "/dotnet/standard/linq/",
             "redirect_document_id": true
+        },
+        {
+            "source_path_from_root": "/docs/standard/native-interop/cross-platform.md",
+            "redirect_url": "/dotnet/standard/native-interop/pinvoke/native-library-loading",
```
Comment (jkotas) [inline .openpublishing.redirection.standard.json:814]:
```suggestion
            "redirect_url": "/dotnet/standard/native-interop/native-library-loading",
```

**Pull request merged**